src/main/scala/util/CreditedIO.scala

```diff
@@ -0,0 +1,152 @@
+
+  def +(that: CreditedDelay): CreditedDelay =
+    CreditedDelay(debit + that.debit, credit + that.credit)
+
+  override def toString = s"${debit}:${credit}"
+}
+
+/** CreditedIO provides credit-debit-based flow control for arbitrary Data.
+  * The sender may only transmit Data when it has non-zero credits.
+  * Receivers provide 0 or 1 credits to senders using the credit field.
+  * Senders consume 0 or 1 credits by setting the debit field.
+  * The bits Data field is DontCare when debit=0.
+  * credit MAY depend combinationally on debit.
+  * debit MAY depend combinationally on credit.
+  * WARNING: The user must ensure the round trip time is > 0.
+  * Failure to comply will result in a combinational loop!
+  */
+final class CreditedIO[T <: Data](gen: T) extends Bundle
+{
+  override def cloneType: this.type = new CreditedIO(genType).asInstanceOf[this.type]
+  def genType: T = gen
+
+  val credit = Input (Bool()) // 1: a credit is given to the sender by the receiver
+  val debit  = Output(Bool()) // 1: a credit is consumed by the sender to transfer 'bits'
+  val bits   = Output(genType)
+
+  /** Provide a DecoupledIO interface for sending CreditedIO[Data].
+    * Convert an IrrevocableIO input to DecoupledIO via Decoupled().
+    * depth controls the maximum number of Data beats inflight.
+    * Sender powers on with credits=depth, so sender and receiver must agree on depth.
+    * pipe=false increases the receiver=>sender trip time by one cycle.
+    * pipe=true causes debit to depend on credit.
+    */
+  def toSender(depth: Int, pipe: Boolean = true): DecoupledIO[T] = {
+    require (depth >= 1)
+    val res = Wire(DecoupledIO(genType))
+    val counter = new CreditedIOCounter(depth, depth)
+    counter.update(this)
+    res.ready := !counter.empty || (pipe.B && credit)
+    debit := res.fire()
+    bits  := res.bits
+    res
+  }
+
+  /** Provide an IrrevocableIO interface for receiving CreditedIO[Data].
+    * Conversion to DecoupledIO is done via application of Decoupled().
+    * depth controls the Queue depth and thus maximum number of elements inflight.
+    * flow=false increases the sender=>receiver trip time by one cycle.
+    * flow=true causes credit to depend on debit.
+    */
+  def toReceiver(depth: Int, flow: Boolean = true): IrrevocableIO[T] = {
+    require (depth >= 1)
+    val enq = Wire(DecoupledIO(genType))
+    enq.valid := debit
+    enq.bits := bits
+    assert (!enq.valid || enq.ready)
+    val res = Queue.irrevocable(enq, depth, pipe=true, flow=flow)
+    credit := res.fire()
+    res
+  }
+
+  /** Add register stages to the sender and receiver paths.
+    * Apply this method to the producer/sender-facing bundle.
+    * The round-trip-time (RTT) is increased by sender+receiver.
+    */
+  def pipeline(debitDelay: Int, creditDelay: Int): CreditedIO[T] = {
+    val res = Wire(CreditedIO(genType))
+    if (debitDelay <= 0) {
+      credit := ShiftRegister(res.credit, creditDelay)
```

Comment:
If this also needs to be reset, you can do `Pipe(res.credit, false.B, creditDelay).valid`, which generates a shfit register whose contents is reset.

Reply:
I just used ShiftRegister's extra args. Seems more readable than creating a Valid chain.

Reply:
agreed, forgot there was such a thing